### PR TITLE
added ValidDurationTicks property to SystemUserTokenKey

### PR DESCRIPTION
**Pull request by diliger**
but wrapper TimeSpan? ValidDuration still doesn't let application to
start.
what am I doing wrong?

changed SystemUser.Static to validate token with ValidDurationTicks as
well, if it HasValue.

https://github.com/StarcounterPrefabs/SignIn/issues/44

### CODE REVIEW

File: Ring3/User/SystemUser.Static.cs [hunk 1]
```diff
@@ -321,7 +325,7 @@ static bool ValidateAuthToken(SystemUserTokenKey token)
             // TODO: Check if token should expire (to old for reuse)?
             TimeSpan ts = new TimeSpan(DateTime.UtcNow.Ticks - token.LastUsed.Ticks);
 
-            if (token.User == null || ts.TotalDays > 7)
+            if (token.User == null || token.Expires < DateTime.UtcNow || token.ValidDurationTicks.HasValue && ts >= TimeSpan.FromTicks(token.ValidDurationTicks.Value))
```
Comment (warpech) [inline Ring3/User/SystemUser.Static.cs]: Wouldn't `if (token.User == null || token.Expires < DateTime.UtcNow)` be enough?

File: Ring5/User/SystemUserTokenKey.cs [hunk 2]
```diff
@@ -1,11 +1,38 @@
...
+        //        if (ValidDurationTicks.HasValue)
+        //            return new TimeSpan(ValidDurationTicks.Value);
+        //        return null;
+        //    }
+        //    set
+        //    {
+        //        ValidDurationTicks = value?.Ticks;
+        //    }
+        //}
     }
```
Comment (warpech) [inline Ring5/User/SystemUserTokenKey.cs]: Can this comment be removed?

File: Ring3/User/SystemUser.Static.cs [hunk 3]
```diff
@@ -325,7 +325,7 @@ static bool ValidateAuthToken(SystemUserTokenKey token)
             // TODO: Check if token should expire (to old for reuse)?
             TimeSpan ts = new TimeSpan(DateTime.UtcNow.Ticks - token.LastUsed.Ticks);
```
Comment (warpech) [inline Ring3/User/SystemUser.Static.cs]: After this change, `ts` is not used.

File: Ring3/User/SystemUser.Static.cs [hunk 4]
```diff
@@ -13,6 +13,8 @@ namespace Simplified.Ring3
 {
     public partial class SystemUser
     {
+        const int tokenValidDays = 7;
```
Comment (warpech) [inline Ring3/User/SystemUser.Static.cs]: This does not follow the .Net convention (fields should use PascalCasing).

Comment (diliger) [inline Ring3/User/SystemUser.Static.cs]: i better make it `protected static`

File: Ring3/User/SystemUser.Static.cs [hunk 5]
```diff
@@ -13,7 +13,7 @@ namespace Simplified.Ring3
 {
     public partial class SystemUser
     {
-        const int tokenValidDays = 7;
+        protected static int tokenValidDays = 7;
```
Comment (warpech) [inline Ring3/User/SystemUser.Static.cs]: The name `tokenValidDays` still does not follow the .Net convention (fields should use PascalCasing).

**Comment by warpech** [comment]
LGTM.

**Comment by warpech** [comment]
I am merging this, even though the related PR https://github.com/StarcounterPrefabs/SignIn/pull/55 still needs some work.

It is important thought that this change is included in Starcounter 2.3 RC.

**Pull request merged**